// @react-md/autocomplete/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
var AutoComplete_1 = require("./AutoComplete");
Object.defineProperty(exports, "AutoComplete", { enumerable: true, get: function () { return AutoComplete_1.default; } });
var HighlightedResult_1 = require("./HighlightedResult");
Object.defineProperty(exports, "HighlightedResult", { enumerable: true, get: function () { return HighlightedResult_1.default; } });
__exportStar(require("./HighlightedResult"), exports);
var useAutoComplete_1 = require("./useAutoComplete");
Object.defineProperty(exports, "useAutoComplete", { enumerable: true, get: function () { return useAutoComplete_1.default; } });
__exportStar(require("./useAutoComplete"), exports);
var utils_1 = require("./utils");
Object.defineProperty(exports, "isResultOf", { enumerable: true, get: function () { return utils_1.isResultOf; } });
Object.defineProperty(exports, "DEFAULT_GET_RESULT_ID", { enumerable: true, get: function () { return utils_1.getResultId; } });
Object.defineProperty(exports, "DEFAULT_GET_RESULT_LABEL", { enumerable: true, get: function () { return utils_1.getResultLabel; } });
Object.defineProperty(exports, "DEFAULT_GET_RESULT_VALUE", { enumerable: true, get: function () { return utils_1.getResultValue; } });
__exportStar(require("./types"), exports);
